### PR TITLE
Homework06

### PR DESCRIPTION
**Pull request by 9Anpanman**

### CODE REVIEW

File: homework06/hw06/bayes.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            for word in xi.split():
+                self.words_counter[word] += 1
+                self.classified_words[word, yi] += 1
+
+        for c in self.classes:
+            self.classes[c] /= len(X)
+
+        self.d = len(self.words_counter)
+
+    def formula(self, cls: str, word: str) -> float:
```
Comment (Dementiy) [inline homework06/hw06/bayes.py]: Формула чего? Лучше давать названия, которые не требуют дополнительных пояснений, даже, если они получаются длинными, например, `calc_word_prob_for_a_given_class`

File: homework06/hw06/bayes.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        return s.translate(translator)
+
+    X, y = [], []
+
+    for target, msg in data:
+        X.append(msg)
+        y.append(target)
+
+    X = [clean(x).lower() for x in X]
+    X_train, y_train, X_test, y_test = X[:3900], y[:3900], X[3900:], y[3900:]
```
Comment (Dementiy) [inline homework06/hw06/bayes.py]: Вот для таких кейсов лучше создавать тесты, а если это должен быть пример использования, то отдельную папочку `examples`

File: homework06/hw06/database.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    label = Column(String)
+
+
+@tp.no_type_check
+def get_session(engine: Engine) -> Session:
+    local_session.configure(bind=engine)
+    return local_session()
+
+
+def table_news(session: Session, news: tp.List[tp.Dict[str, tp.Union[int, str]]]) -> None:
```
Comment (Dementiy) [inline homework06/hw06/database.py]: Что делает `table_news`? Имена функция это глаголы, классов - существительные.
```python
class NewsType(tp.TypedDict):
    title: str
    author: str
    url: str
    points: int

def insert_into_news_table(session: Session, raw_news_list: tp.List[NewsType]) -> None: 
    news_list = [News(**news_data) for news_data in raw_news_list]
    session.add_many(news_list)
    session.commit()
```

File: homework06/hw06/database.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            url=news[i]["url"],
+            points=news[i]["points"],
+        )
+        session.add(means)
+    session.commit()
+
+
+@tp.no_type_check
+def change_label(session: Session, data: tp.List[tp.Union[str, int]]) -> None:
+    item = session.query(News).get(data[1])
```
Comment (Dementiy) [inline homework06/hw06/database.py]: А что такое data[1], из контекста можно понять, что data[0] это label. Как понять при вызове функции change_label, что нужно передать в качестве аргументов?

File: homework06/hw06/database.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    item.label = data[0]
+    session.commit()
+
+
+def next_news(session: Session) -> None:
+    news = get_news(url="https://news.ycombinator.com/newest")
+    news_news = []
```
Comment (Dementiy) [inline homework06/hw06/database.py:53]: `unseen_news`, `new_news`?

File: homework06/hw06/database.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    for something in news:
+        main, name = something["title"], something["author"]
+        if not list(session.query(News).filter(News.title == main, News.author == name)):
+            news_news.append(something)
+    table_news(session, news_news)
```
Comment (Dementiy) [inline homework06/hw06/database.py]: Про эту функцию см. выше

File: homework06/hw06/database.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    item.label = data[0]
+    session.commit()
+
+
+def next_news(session: Session) -> None:
+    news = get_news(url="https://news.ycombinator.com/newest")
+    news_news = []
+    for something in news:
+        main, name = something["title"], something["author"]
+        if not list(session.query(News).filter(News.title == main, News.author == name)):
```
Comment (Dementiy) [inline homework06/hw06/database.py:56]: У query есть метод `exists`

File: homework06/hw06/database.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+@tp.no_type_check
+def change_label(session: Session, data: tp.List[tp.Union[str, int]]) -> None:
+    item = session.query(News).get(data[1])
+    item.label = data[0]
+    session.commit()
+
+
+def next_news(session: Session) -> None:
+    news = get_news(url="https://news.ycombinator.com/newest")
```
Comment (Dementiy) [inline homework06/hw06/database.py]: Неясно из названия, что эта функция делает. И нужно url указать в качестве аргумента.
```python
def get_new_news(session: Session, url: str) -> tp.List[NewsType]:
    # ...

# и в месте вызова
new_news = get_new_news(session, url="https://news.ycombinator.com/newest")
insert_into_news_table(new_news)
```
Теперь получение новых новостей можно протестировать отдельно.

File: homework06/hw06/hackernews.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    label, id = label.split("=")[1], int(id.split("=")[1])
+    change_label(s, [label, id])
+    redirect("/news")
+
+
+@tp.no_type_check
+@route("/update")
+def update_news():
+    s = get_session(engine)
+    next_news(s)
```
Comment (Dementiy) [inline homework06/hw06/hackernews.py]: См. комментарий выше

File: homework06/hw06/hackernews.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    return template("news_template", rows=rows)
+
+
+@tp.no_type_check
+@route("/add_label/")
+def add_label():
+    s = get_session(engine)
+    http_request = request.query_string
+    label, id = http_request.split("&")
+    label, id = label.split("=")[1], int(id.split("=")[1])
```
Comment (Dementiy) [inline homework06/hw06/hackernews.py]: Из запроса сразу можно получить нужные параметры, смотрите документацию bottle

File: homework06/hw06/hackernews.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+
+@tp.no_type_check
+@route("/classify")
+def classify_news():
+    s = get_session(engine)
+    model = NaiveBayesClassifier()
+    train_set = s.query(News).filter(News.label != None).all()
+    model.fit([clean(news.title).lower() for news in train_set], [news.label for news in train_set])
+    test = s.query(News).filter(News.label == None).all()
```
Comment (Dementiy) [inline homework06/hw06/hackernews.py:50]: Подготовку данных вынести надо:

```python
@route("/classify")
def classify_news():
    s = get_session(engine)
    model = NaiveBayesClassifier()
    train_x, train_y, test_x = get_train_test_set(s)
    model.fit(train_x, train_y)
    predictions = model.predict(test_x)
    return template(....)
```

File: homework06/hw06/hackernews.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    cell = list(map(lambda x: model.predict(x.title), test))
+    return template(
+        "color_template", rows=list(map(lambda x: (x[1], colors[cell[x[0]]]), enumerate(test)))
```
Comment (Dementiy) [inline homework06/hw06/hackernews.py:53]: Нечитаемый кусок

File: homework06/hw06/scraputils.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+    title = parser.select(".storylink")
+    points = parser.select(".score")
+    subtext = parser.select(".subtext")
+
+    for i in title:
+        title_list.append(i.text)
+        link = i.get("href", None)
+        if link.startswith("item"):
+            links_list.append("https://news.ycombinator.com/" + link)
```
Comment (Dementiy) [inline homework06/hw06/scraputils.py]: Ссылочку либо аргументом делать, либо константой

**Review by Dementiy — CHANGES_REQUESTED** [review]

File: homework06/hw06/database.py [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+def make_table_news(session: Session, news: tp.List[tp.Dict[str, tp.Union[int, str]]]) -> None:
+    for i in range(len(news)):
+        means = News(
+            title=news[i]["title"],
+            author=news[i]["author"],
+            url=news[i]["url"],
+            points=news[i]["points"],
+        )
+        session.add(means)
```
Comment (Dementiy) [inline homework06/hw06/database.py:40]: `add_many` можно было использовать

**Review by Dementiy — APPROVED** [review]
В целом хорошо получилось, принято
